models.rb

```diff
@@ -0,0 +1,101 @@
+require 'rubygems'
+require 'data_mapper'
+require 'dm-validations'
+
+# A User can be an owner or requester
+class User
+  include DataMapper::Resource
+
+  property :id, Serial
+  property :username, String, required: true
+  property :email, String,
+           format: :email_address,
+           required: true,
+           unique: true,
+           messages: {
+             format: 'You must enter a valid email address.'
+           }
+
+  property :password, BCryptHash,
+           required: true
+
+  attr_accessor :password_confirmation
+  validates_confirmation_of :password
+  validates_length_of :password_confirmation,
+                      min: 6,
+                      if: :password_required?
+
+  has n, :stalls, child_key: [:owner_id]
+  has n, :rental_requests, child_key: [:requester_id]
+
+  def password_required?
+    self.new? || self.dirty?
+  end
+
+  def valid_password?(unhashed_password)
+    password == unhashed_password
+  end
+
+  def self.find_by_credentials(email, password)
+    user = User.first(email: email) || User.new(email: email)
+    unless user.valid_password?(password)
+      user.errors.add(:general, "We couldn't find an account with that email \
+        and password. Can you double-check them?")
+    end
+    user
+  end
+end
+
+# Represents a horse stall that can be requested
+class Stall
+  include DataMapper::Resource
+
+  property :id, Serial
+  property :title, String, required: true, unique: true
+  property :description, Text, required: true
+  property :city, String, required: true
+  property :state, String, required: true
+  property :zip, String, required: true
+
+  belongs_to :owner, 'User',
+             child_key: [:owner_id],
+             required: true
+  has n, :rental_requests
+end
+
+# Represents a user's request of an owner's stall
+class RentalRequest
+  include DataMapper::Resource
+
+  property :id, Serial
+  property :date, Date, required: true
+  property :status, Enum[:accepted, :pending, :declined, :canceled], required: true
+  property :message, Text
+
+  ### WHY IS THIS ADDING :valid_date?=>[nil] TO ERRORS?
```

Comment:
I don't know.  Are you sure you're supposed to `return false` from a custom validation?  The example on the DataMapper website doesn't work that way.  I'm not sure what the consequences are.